pkg/cpio/types.go

```diff
@@ -33,27 +34,61 @@ type RecordFormat interface {
 	Writer(w io.Writer) RecordWriter
 }
 
-func StaticRecord(contents string, info Info) Record {
+func StaticRecord(contents []byte, info Info) Record {
 	info.FileSize = uint64(len(contents))
 	return Record{
-		bytes.NewReader([]byte(contents)),
-		info,
+		ReadCloser: NewBytesReadCloser(contents),
+		Info:       info,
 	}
 }
 
-type eofReader struct{}
+type ReadCloser struct {
+	io.Reader
+}
 
-func (eofReader) Read(p []byte) (int, error) {
-	return 0, io.EOF
+func (b ReadCloser) Close() error {
+	return nil
+}
+
+func NewBytesReadCloser(contents []byte) io.ReadCloser {
+	return &ReadCloser{bytes.NewReader(contents)}
+}
+
+func NewReadCloser(r io.Reader) io.ReadCloser {
+	return &ReadCloser{r}
 }
 
 func EmptyRecord(info Info) Record {
 	return Record{
-		eofReader{},
+		NewBytesReadCloser([]byte{}),
 		info,
 	}
 }
 
+type DeferReader struct {
+	Name string
+	File *os.File
+}
+
+func (r *DeferReader) Read(p []byte) (int, error) {
+	if r.File == nil {
+		f, err := os.Open(r.Name)
+		if err != nil {
+			return -1, err
+		}
+		r.File = f
+	}
+	return r.File.Read(p)
+}
+
+func (r *DeferReader) Close() error {
+	return r.File.Close()
```

Comment:
ah, OK, I misunderstood where you were going there. I'm not convinced this matters that much but I'll look at it. The reason I'm not convinced is that cpio is a single-pass thing -- if they're calling Close(), they're done, and the only real reason to return an error is so they know. This file won't be touched again. This class is kind of specific to cpio and that's what makes it simple. 

Come to think of it ... I believe if you call f.Close and try to read from f again you get the correct error, which is that the file is close. If so, that argues for calling close and doing nothing else, as repeated reads will return EOF. I will look again, but that's certainly the right thing to do.